File: app/src/main/java/com/ninpou/qbits/request/RequestFragment.java
package com.ninpou.qbits.request;

import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebSettings;
import android.widget.AdapterView;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.constraintlayout.widget.Group;
import androidx.fragment.app.Fragment;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.ninpou.qbits.R;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import top.srsea.lever.ui.Toasts;


public class RequestFragment extends Fragment {
    private static final String URL_REGEX = "(http|https)://[-A-Za-z0-9+&@#/%?=~_|!:,.;]+[-A-Za-z0-9+&@#/%=~_|]";
    private static final String KEY_USER_AGENT = "http.agent";
    private static final String HEADER_KEY_USER_AGENT = "User-Agent";
    private static final String HEADER_KEY_CONTENT_TYPE = "Content-Type";
    private static final String DEFAULT_CONTENT_TYPE = "application/json";
    private static final String BUNDLE_KEY_RESPONSE = "response_key";

    private OkHttpClient client = new OkHttpClient();
    private Handler handler = new Handler();
    private List<TextInputLayout> headers = new ArrayList<>();
    private ProgressDialog progressDialog;
    private FloatingActionButton sendButton;
    private Group bodyGroup;
    private TextInputEditText bodyEdit;
    private TextInputEditText urlEdit;
    private Spinner functionSpinner;
    private TextInputEditText userAgentEdit;
    private TextInputEditText contentTypeEdit;
    private LinearLayout headersLayout;
    private TextView resetButton;
    private TextView addButton;
    private HttpMethod method = HttpMethod.GET;

    public static RequestFragment newInstance() {
        return new RequestFragment();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View rootView = inflater.inflate(R.layout.fragment_request, container, false);
        initView(rootView);
        return rootView;
    }

    private void initView(View root) {
        sendButton = root.findViewById(R.id.button_send);
        urlEdit = root.findViewById(R.id.et_url);
        functionSpinner = root.findViewById(R.id.spinner);
        userAgentEdit = root.findViewById(R.id.et_user_agent);
        contentTypeEdit = root.findViewById(R.id.et_content_type);
        bodyGroup = root.findViewById(R.id.group_body);
        bodyEdit = root.findViewById(R.id.et_body);
        addButton = root.findViewById(R.id.btn_add);
        resetButton = root.findViewById(R.id.btn_reset);
        headersLayout = root.findViewById(R.id.ll_headers);
        progressDialog = new ProgressDialog(getActivity());
        progressDialog.setMessage(getString(R.string.load));
        progressDialog.setCanceledOnTouchOutside(false);
        String userAgent;
        try {
            userAgent = WebSettings.getDefaultUserAgent(getActivity());
        } catch (Exception e) {
            userAgent = System.getProperty(KEY_USER_AGENT);
        }
        userAgentEdit.setText(userAgent);
        contentTypeEdit.setText(DEFAULT_CONTENT_TYPE);
        sendButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                sendRequest();
            }
        });
        functionSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                method = HttpMethod.values()[position];
                if (method == HttpMethod.GET) {
                    bodyGroup.setVisibility(View.INVISIBLE);
                } else {
                    bodyGroup.setVisibility(View.VISIBLE);
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
            }
        });
        addButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                addHeader();
            }
        });
        resetButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                for (View view : headers) {
                    headersLayout.removeView(view);
                }
                headers.clear();
            }
        });
    }

    private void addHeader() {
        final View view = LayoutInflater.from(requireContext())
                .inflate(R.layout.alert_add_header, null);
        new AlertDialog.Builder(requireContext())
                .setTitle(R.string.add_alert_title)
                .setView(view)
                .setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        EditText editText = view.findViewById(R.id.et_add_header);
                        String key = editText.getText().toString();
                        if (key.isEmpty()) return;
                        TextInputLayout layout = new TextInputLayout(requireContext());
                        TextInputEditText header = new TextInputEditText(requireContext());
                        header.setHint(key);
                        layout.addView(header);
                        headersLayout.addView(layout);
                        headers.add(layout);
                    }
                })
                .setNegativeButton(R.string.cancle, null)
                .show();
    }

    private Request initRequest() {
        String url = urlEdit.getText().toString();
        String userAgent = userAgentEdit.getText().toString();
        String contentType = contentTypeEdit.getText().toString();
        if (url.isEmpty()) {
            Toasts.of(R.string.url_empty_tip).showShort();
            return null;
        }
        if (!url.matches(URL_REGEX)) {
            Toasts.of(R.string.url_illegal_tip).showShort();
            return null;
        }
        Request.Builder builder = new Request.Builder();
        builder.url(url);
        try {
            if (!userAgent.isEmpty())
                builder.addHeader(HEADER_KEY_USER_AGENT, userAgent);
            if (!contentType.isEmpty())
                builder.addHeader(HEADER_KEY_CONTENT_TYPE, contentType);
            for (TextInputLayout header : headers) {
                if (header.getEditText() == null || header.getEditText().getText() == null
                        || header.getEditText().getText().toString().isEmpty())
                    continue;
                EditText editText = header.getEditText();
                builder.addHeader(editText.getHint().toString(), editText.getText().toString());
            }
        } catch (IllegalArgumentException e) {
            Toasts.of(R.string.illegal_header_tip).showShort();
            return null;
        }
        String bodyStr = bodyEdit.getText().toString();
        RequestBody body = RequestBody.create(MediaType.parse(contentType), bodyStr);
        switch (method) {
            case GET:
                builder.get();
                break;
            case POST:
                builder.post(body);
                break;
            case HEAD:
                builder.head();
                break;
            case PUT:
                builder.put(body);
                break;
            case DELETE:
                builder.delete(body);
                break;
        }
        return builder.build();
    }

    private void sendRequest() {
        Request request = initRequest();
        if (request == null) return;
        progressDialog.show();
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                final String message = e.getMessage();
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        progressDialog.cancel();
                        Toasts.of(R.string.request_fail_tip).showShort();
                    }
                });
            }

            @Override
            public void onResponse(@NonNull Call call, @NonNull final Response response)
                    throws IOException {
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        Intent intent = new Intent(getActivity(), ResponseActivity.class);
                        ResponseInfo info = new ResponseInfo(response);
                        progressDialog.cancel();
                        intent.putExtra(BUNDLE_KEY_RESPONSE, info);
                        startActivity(intent);
                    }
                });
            }
        });
    }

    private enum HttpMethod {
        GET, POST, HEAD, PUT, DELETE
    }
}

File: app/src/main/java/com/ninpou/qbits/request/RequestFragment.kt
package com.ninpou.qbits.request

import android.app.ProgressDialog
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.WebSettings
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.EditText
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import com.ninpou.qbits.R
import com.ninpou.qbits.util.*
import kotlinx.android.synthetic.main.fragment_request.*
import kotlinx.android.synthetic.main.fragment_request.view.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import okhttp3.*
import java.io.IOException
import java.util.*

class RequestFragment : Fragment() {
    private var client = OkHttpClient()
    private val headers: MutableList<TextInputLayout> = ArrayList()
    private var progressDialog: ProgressDialog? = null
    private var method = HttpMethod.GET
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val rootView = inflater.inflate(R.layout.fragment_request, container, false)
        initView(rootView)
        return rootView
    }

    private fun initView(root: View) {
        progressDialog = ProgressDialog(APP_ACTIVITY)
        progressDialog?.setMessage(getString(R.string.load))
        progressDialog?.setCanceledOnTouchOutside(false)
        val userAgent: String? = try {
            WebSettings.getDefaultUserAgent(APP_ACTIVITY)
        } catch (e: Exception) {
            System.getProperty(KEY_USER_AGENT)
        }
        root.et_user_agent.setText(userAgent)
        root.et_content_type.setText(DEFAULT_CONTENT_TYPE)
        root.button_send.setOnClickListener { sendRequest() }
        root.spinner.onItemSelectedListener = object : OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View, position: Int, id: Long) {
                method = HttpMethod.values()[position]
                if (method == HttpMethod.GET) {
                    group_body.visibility = View.INVISIBLE
                } else {
                    showViews(group_body)
                }
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }
        root.btn_add.setOnClickListener { addHeader() }
        root.btn_reset.setOnClickListener {
            for (view in headers) {
                ll_headers.removeView(view)
            }
            headers.clear()
        }
    }

    private fun addHeader() {
        val view = LayoutInflater.from(requireContext())
                .inflate(R.layout.alert_add_header, null)
        AlertDialog.Builder(requireContext())
                .setTitle(R.string.add_alert_title)
                .setView(view)
                .setPositiveButton(R.string.ok, DialogInterface.OnClickListener { _, _ ->
                    val editText = view.findViewById<EditText>(R.id.et_add_header)
                    val key = editText.text.toString()
                    if (key.isEmpty()) return@OnClickListener
                    val layout = TextInputLayout(requireContext())
                    val header = TextInputEditText(requireContext())
                    header.hint = key
                    layout.addView(header)
                    ll_headers.addView(layout)
                    headers.add(layout)
                })
                .setNegativeButton(R.string.cancel, null)
                .show()
    }

    private fun initRequest(): Request? {
        val url = et_url.text.toString()
        val userAgent = et_user_agent.text.toString()
        val contentType = et_content_type.text.toString()
        if (url.isEmpty()) {
            showShortToast(getString(R.string.url_empty_tip))
            return null
        }
        if (!url.matches(URL_REGEX.toRegex())) {
            showShortToast(getString(R.string.url_illegal_tip))
            return null
        }
        val builder = Request.Builder()
        builder.url(url)
        try {
            if (userAgent.isNotEmpty()) builder.addHeader(HEADER_KEY_USER_AGENT, userAgent)
            if (contentType.isNotEmpty()) builder.addHeader(HEADER_KEY_CONTENT_TYPE, contentType)
            for (header in headers) {
                if (header.editText == null || header.editText?.text == null || header.editText?.text.toString().isEmpty()) continue
                val editText = header.editText
                builder.addHeader(editText?.hint.toString(), editText?.text.toString())
            }
        } catch (e: IllegalArgumentException) {
            showShortToast(getString(R.string.illegal_header_tip))
            return null
        }
        val bodyStr = Objects.requireNonNull(et_body.text).toString()
        val body = RequestBody.create(MediaType.parse(contentType), bodyStr)
        when (method) {
            HttpMethod.GET -> builder.get()
            HttpMethod.POST -> builder.post(body)
            HttpMethod.HEAD -> builder.head()
            HttpMethod.PUT -> builder.put(body)
            HttpMethod.DELETE -> builder.delete(body)
        }
        return builder.build()
    }

    private fun sendRequest() {
        val request = initRequest() ?: return
        progressDialog?.show()
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                val message = e.message
                GlobalScope.launch(Dispatchers.Main) {
                    progressDialog?.cancel()
                    showShortToast(getString(R.string.request_fail_tip))
                }
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                GlobalScope.launch(Dispatchers.Main) {
                    val intent = Intent(APP_ACTIVITY, ResponseActivity::class.java)
                    val info = ResponseInfo(response)
                    progressDialog?.cancel()
                    intent.putExtra(BUNDLE_KEY_RESPONSE, info)
                    startActivity(intent)
                }
            }
        })
    }

    private enum class HttpMethod {
        GET, POST, HEAD, PUT, DELETE
    }

    companion object {
        fun newInstance(): RequestFragment {
            return RequestFragment()
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        progressDialog = null
    }
}

